### PR TITLE
Make all communication timeouts configurable.

### PR DESCRIPTION
**Pull request by smadala**
## Description
- There are various timeout error on slow machines. This PR address to make all the timeout 
configurable.
- Refactor DefaultEngineInvoker.cs and Datacollector Program.cs to make them testable.
## Related issue
 #1544 #1361

**Exxample output**

```
PS C:\Users\samadala\src\vstest> $env:VSTEST_CONNECTION_TIMEOUT=1
PS C:\Users\samadala\src\vstest> C:\Users\samadala\src\vstest\artifacts\Debug\net451\win7-x64\vstest.console.exe C:\Users\samadala\src\vstest\test\TestAssets\SimpleTestProject\bin\Debug\net451\SimpleTestProject.dll /Inisolation
Microsoft (R) Test Execution Command Line Tool Version 15.7.0-dev
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: vstest.console process failed to connect to testhost process after 1 seconds. This may occur due to machine slowness, please set environment variable VSTEST_CONNECTION_TIMEOUT to increase timeout.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.HandleConnectionFailure(Boolean connected, Int32 connTimeout)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyExecutionManager.StartTestRun(TestRunCriteria testRunCriteria, ITestRunEventsHandler eventHandler)
```

### CODE REVIEW

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformEnvironment.cs [hunk 1]
```diff
@@ -61,5 +61,18 @@ public int GetCurrentManagedThreadId()
...
+        {
+            var isValueSet = false;
+            var envVarValue = Environment.GetEnvironmentVariable(envVar);
+            if (!string.IsNullOrEmpty(envVarValue) && double.TryParse(envVarValue, out value))
+            {
+                isValueSet = true;
+            }
+
+            return isValueSet;
+        }
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformEnvironment.cs]: Not required in Abstraction. 
For testing, either have EnvironmentHelper or directory use it in the tests.

File: src/Microsoft.TestPlatform.ObjectModel/Constants.cs [hunk 2]
```diff
@@ -162,6 +162,11 @@ public static class Constants
         /// </summary>
         public const string ResultsDirectoryName = "TestResults";
 
+        /// <summary>
+        /// Environment variable name to increase communication timeouts between processes.
+        /// </summary>
+        public const string VstestTimeoutIncreaseByTimes = "VSTEST_TIMEOUT_INCREASE_BY_TIMES";
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.ObjectModel/Constants.cs]: >VSTEST_TIMEOUT_INCREASE_BY_TIMES [](start = 60, length = 32)

How about VSTEST_CONNECTION_TIMEOUT?
We should have a simple settings, also make it consistent across

File: src/datacollector/Resources/xlf/Resources.cs.xlf [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        <count count-type="x-match99-90%">0</count>
+        <count count-type="x-match89-75%">0</count>
+        <count count-type="x-mt">27</count>
+        <count count-type="repetition">0</count>
+      </count-group>
+    </header>
+    <body>
+      <trans-unit id="DataCollectorFailedToConnetToVSTestConsole">
+        <source>Datacollector process failed to connect to vstest.console in {0} milliseconds, Set enviroment variable {1} to increase timeout.</source>
+        <target state="new">Datacollector process failed to connect to vstest.console in {0} milliseconds, Set enviroment variable {1} to increase timeout.</target>
```
Comment (singhsarab) [inline src/datacollector/Resources/xlf/Resources.cs.xlf]: >Datacollector process failed to connect to vstest.console in {0} milliseconds [](start = 28, length = 77)

Datacollector process failed to connect to vstest.console in {0} milliseconds. Please set environment variable {1} to increase the connection timeout.

File: src/datacollector/DataCollectionMain.cs [hunk 4]
```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mayankbansal018) [inline src/datacollector/DataCollectionMain.cs]: Why did we create a new class?

File: src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformEnvironment.cs (hunk 1 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformEnvironment.cs]: Done.

File: src/Microsoft.TestPlatform.ObjectModel/Constants.cs (hunk 2 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.ObjectModel/Constants.cs]: Makes sense, Done.

File: src/datacollector/Resources/xlf/Resources.cs.xlf (hunk 3 above)
Comment (smadala) [reply src/datacollector/Resources/xlf/Resources.cs.xlf]: @singhsarab @mayankbansal018 Agreed. Done.

File: src/datacollector/DataCollectionMain.cs (hunk 4 above)
Comment (smadala) [reply src/datacollector/DataCollectionMain.cs]: For testing refactored it.

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: do we want to add more in CoreUtilities dll? can we move this to Testplatform.Utilities?

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CoreUtilities.Helpers
+{
+    using System;
+    using ObjectModel;
+
+    public class EnvironmentHelper
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: there is an IEnvironment Interface in platform abstractions, should it not move there?

File: src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs [hunk 7]
```diff
@@ -207,20 +206,22 @@ public void Initialize()
             this.dataCollectionProcessId = this.dataCollectionLauncher.LaunchDataCollector(null, this.GetCommandLineArguments(this.dataCollectionPort));
             EqtTrace.Info("ProxyDataCollectionManager.Initialize: Launched datacollector processId: {0} port: {1}", this.dataCollectionProcessId, this.dataCollectionPort);
 
-            ChangeConnectionTimeoutIfRequired(dataCollectionProcessId);
+            var connectionTimeout = GetConnectionTimeout(dataCollectionProcessId);
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/ProxyDataCollectionManager.cs]: nit: please use this, just to follow current coding convention

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 8]
```diff
@@ -49,21 +48,38 @@ public abstract class ProxyOperationManager
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProxyOperationManager"/> class
+        /// </summary>
+        /// <param name="requestData"></param>
         /// <param name="requestSender">Request Sender instance.</param>
         /// <param name="testHostManager">Test host manager instance.</param>
-        /// <param name="clientConnectionTimeout">Client Connection Timeout.</param>
-        protected ProxyOperationManager(IRequestData requestData, ITestRequestSender requestSender, ITestRuntimeProvider testHostManager, int clientConnectionTimeout)
+        /// <param name="processHelper">IProcessHelper implementation. </param>
+        internal ProxyOperationManager(IRequestData requestData,
```
Comment (singhsarab) [inline src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: remove this.

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyOperationManagerTests.cs [hunk 9]
```diff
@@ -204,24 +209,24 @@ public void SetupChannelShouldNotWaitForTestHostConnectionIfConnectionIsInitiali
...
-            Environment.SetEnvironmentVariable("VSTEST_CONNECTION_TIMEOUT", "100");
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, "100");
 
             this.mockRequestSender.Setup(rs => rs.WaitForRequestHandlerConnection(100000)).Returns(true);
             this.testOperationManager.SetupChannel(Enumerable.Empty<string>(), CancellationToken.None);
 
             this.mockRequestSender.Verify(rs => rs.WaitForRequestHandlerConnection(100000), Times.Exactly(1));
 
-            this.connectionTimeout = 400;
+            this.connectionTimeout = EnvironmentHelper.DefaultConnectionTimeout;
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyOperationManagerTests.cs]: remove this

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyOperationManagerTests.cs [hunk 10]
```diff
@@ -204,24 +209,24 @@ public void SetupChannelShouldNotWaitForTestHostConnectionIfConnectionIsInitiali
...
         [TestMethod]
         public void SetupChannelShouldNotHonorGarbageTimeOutSetByUser()
         {
-            Environment.SetEnvironmentVariable("VSTEST_CONNECTION_TIMEOUT", "garbage");
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, "garbage");
 
             this.testOperationManager.SetupChannel(Enumerable.Empty<string>(), CancellationToken.None);
 
-            this.mockRequestSender.Verify(rs => rs.WaitForRequestHandlerConnection(this.connectionTimeout), Times.Exactly(1));
+            this.mockRequestSender.Verify(rs => rs.WaitForRequestHandlerConnection(this.connectionTimeout));
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/Client/ProxyOperationManagerTests.cs]: revert this exactly change

File: test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs [hunk 11]
```diff
@@ -69,11 +70,11 @@ public void InitializeShouldSetTimeoutBasedOnTimeoutEnvironmentVarible()
         {
 
             var timeout = 10;
-            Environment.SetEnvironmentVariable(ProxyDataCollectionManager.TimeoutEnvironmentVaribleName, timeout.ToString());
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, timeout.ToString());
             this.mockDataCollectionRequestSender.Setup(x => x.WaitForRequestHandlerConnection(timeout * 1000)).Returns(true);
 
             this.proxyDataCollectionManager.Initialize();
-            Environment.SetEnvironmentVariable(ProxyDataCollectionManager.TimeoutEnvironmentVaribleName, string.Empty);
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, string.Empty);
```
Comment (singhsarab) [inline test/Microsoft.TestPlatform.CrossPlatEngine.UnitTests/DataCollection/ProxyDataCollectionManagerTests.cs]: Cleanup?

File: test/datacollector.UnitTests/DataCollectorMainTests.cs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        [TestCleanup]
+        public void CleanUp()
+        {
+            Environment.SetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout, "");
+        }
+
+        [TestMethod]
+        public void RunShouldTimeoutBasedOneEnvVariable()
```
Comment (singhsarab) [inline test/datacollector.UnitTests/DataCollectorMainTests.cs]: On*

File: test/datacollector.UnitTests/DataCollectorMainTests.cs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        }
+
+        [TestMethod]
+        public void RunShouldThrowIfTimeoutOccured()
+        {
+            this.mockDataCollectionRequestHandler.Setup(rh => rh.WaitForRequestSenderConnection(It.IsAny<int>())).Returns(false);
+            var message = Assert.ThrowsException<TestPlatformException>(() => this.dataCollectorMain.Run(args)).Message;
+            Assert.AreEqual(message,
+                string.Format(
+                    TestPlatform.DataCollector.Resources .Resources.DataCollectorFailedToConnetToVSTestConsole,
```
Comment (singhsarab) [inline test/datacollector.UnitTests/DataCollectorMainTests.cs]: Nit: Connect*

File: test/datacollector.UnitTests/DataCollectorMainTests.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+    [TestClass]
+    public class DataCollectorMainTests
+    {
+        private readonly string[] args = {"--port", "1025", "--parentprocessid", "100" };
+        private Mock<IProcessHelper> mockProcessHelper;
+        private Mock<IEnvironment> mockEnvironment;
+        private Mock<IDataCollectionRequestHandler> mockDataCollectionRequestHandler;
+        private DataCollectorMain dataCollectorMain;
+
```
Comment (singhsarab) [inline test/datacollector.UnitTests/DataCollectorMainTests.cs]: extra line*

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs (hunk 5 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: CoreUtilities have basic utilities on BCL. TestPlatform.Utilities has OM + BCL Utilities.

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs (hunk 6 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: We will go for PlatformAbstration if there is need to do #if defs. This function not required that.

**Comment by smadala** [comment]
@singhsarab @mayankbansal018 Addressed all the comments. Please have a look once more.

File: src/testhost.x86/DefaultEngineInvoker.cs [hunk 15]
```diff
@@ -46,28 +48,157 @@ internal class DefaultEngineInvoker :
...
+ System.IO.File.ReadAllText(AppDomain.CurrentDomain.SetupInformation.ConfigurationFile);
+                EqtTrace.Info("DefaultEngineInvoker: Using Application Configuration: '{0}'", appConfigText);
+#endif
             }
-            else
+
+            this.SetParentProcessExitCallback(argsDictionary);
+
+            this.requestHandler.ConnectionInfo =
+                DefaultEngineInvoker.GetConnectionInfo(out var endpoint, argsDictionary);
```
Comment (mayankbansal018) [inline src/testhost.x86/DefaultEngineInvoker.cs]: >endpoint [](start = 63, length = 8)

remove endpoint, move logging in GetConnectionInfo()

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        public const string VstestConnectionTimeout = "VSTEST_CONNECTION_TIMEOUT";
+        public const int DefaultConnectionTimeout = 90; // seconds
+
+        /// <summary>
+        /// Get timeout based on environment variable VSTEST_CONNECTION_TIMEOUT.
+        /// </summary>
+        public static int GetConnectionTimeout()
+        {
+            var envVarValue = Environment.GetEnvironmentVariable(EnvironmentHelper.VstestConnectionTimeout);
+            if (!string.IsNullOrEmpty(envVarValue) && int.TryParse(envVarValue, out int value))
```
Comment (mayankbansal018) [inline src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: if user sets a very small value say 10, should we honor it? I would say we keep 90 or 60 something as min, & users value will only be honored if it is > min val
More over here we are not even checking for -ve values

File: src/testhost.x86/DefaultEngineInvoker.cs (hunk 15 above)
Comment (smadala) [reply src/testhost.x86/DefaultEngineInvoker.cs]: Done.

File: src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs [hunk 17]
```diff
@@ -151,21 +145,16 @@ public virtual bool SetupChannel(IEnumerable<string> sources, CancellationToken
...
                     // Increase connection timeout when debugging is enabled.
-                    connTimeout = 5 * this.connectionTimeout;
+                    connTimeout *= 5;
                 }
 
                 // Wait for a timeout for the client to connect.
-                if (!this.testHostLaunched || !this.RequestSender.WaitForRequestHandlerConnection(connTimeout))
-                {
-                    var errorMsg = CrossPlatEngineResources.InitializationFailed;
+                var connected = true;
```
[the remark this answers is not in the document]
Comment (smadala) [reply src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs]: Done.

File: src/Microsoft.TestPlatform.TestHostProvider/Hosting/TestHostManagerCallbacks.cs [hunk 18]
```diff
@@ -70,6 +64,15 @@ public static void ErrorReceivedCallback(StringBuilder testHostProcessStdError,
...
+            if (exitCode != 0)
+            {
+                EqtTrace.Error("TestHostManagerCallbacks.ExitCallBack: Testhost processId: {0} exited with exitcode: {1} error: '{2}'", procId, exitCode, testHostProcessStdErrorStr);
+            }
+            else
+            {
+                EqtTrace.Info("TestHostManagerCallbacks.ExitCallBack: Testhost processId: {0} exited with exitcode: 0 error: '{1}'", procId, testHostProcessStdErrorStr);
+            }
+
             onHostExited(new HostProviderEventArgs(testHostProcessStdErrorStr, exitCode, procId));
```
[the remark this answers is not in the document]
Comment (smadala) [reply src/Microsoft.TestPlatform.TestHostProvider/Hosting/TestHostManagerCallbacks.cs]: If exit code is non-zero we are writing to error level.

File: src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs (hunk 16 above)
Comment (smadala) [reply src/Microsoft.TestPlatform.CoreUtilities/Helpers/EnvironmentHelper.cs]: Handled -ve case. Keep min value 1, giving more choice to user.

**Review by mayankbansal018 — APPROVED** [review]

**Pull request merged**